File: statefun-sdk-js/test/types.test.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2014 The Apache Software Foundation

import {describe, expect} from '@jest/globals'
import {StateFun} from '../src/statefun';
import {TypedValueSupport} from "../src/types";
import {Type} from "../src/core";

function roundTrip<T>(tpe: Type<T>, value: T): T {
    const bytes = tpe.serialize(value);
    return tpe.deserialize(bytes);
}

describe('Simple Serialization Test', () => {
    it('should serialize a True booleans', () => {
        let actual = roundTrip(StateFun.booleanType(), true);
        expect(actual).toStrictEqual(true);
    });

    it('should serialize a False booleans', () => {
        let actual = roundTrip(StateFun.booleanType(), false);
        expect(actual).toStrictEqual(false);
    });

    it('should serialize a string', () => {
        let actual = roundTrip(StateFun.stringType(), "Hello world");
        expect(actual).toStrictEqual("Hello world");
    });

    it('should serialize an empty string', () => {
        let actual = roundTrip(StateFun.stringType(), "");
        expect(actual).toStrictEqual("");
    });

    it('should serialize an int', () => {
        let actual = roundTrip(StateFun.intType(), 12345);
        expect(actual).toStrictEqual(12345);
    });

    it('should serialize a float', () => {
        let actual = roundTrip(StateFun.floatType(), 123.0);
        expect(actual).toStrictEqual(123.0);
    });

    it('should serialize Json', () => {
        let actual = roundTrip(StateFun.jsonType("foo/bar"), {a: 1, b: "world"});
        expect(actual).toStrictEqual({a: 1, b: "world"});
    });

    it('should round trip a TypedValue of a string.', () => {
        let box = TypedValueSupport.toTypedValue("hello", StateFun.stringType());
        let got = TypedValueSupport.parseTypedValue(box, StateFun.stringType());

        expect(got).toStrictEqual("hello");
    })

    it('should box a NULL value as a missing value', () => {
        let box = TypedValueSupport.toTypedValue(null, StateFun.stringType());

        expect(box.getHasValue()).toStrictEqual(false);
    })

    it('Should unbox a TypedValue with a missing value as NULL', () => {
        let box = TypedValueSupport.toTypedValue(null, StateFun.stringType());
        let got = TypedValueSupport.parseTypedValue(box, StateFun.stringType());

        expect(got).toStrictEqual(null);
    })

    it('Should fail to unbox with a wrong type', () => {
        let box = TypedValueSupport.toTypedValue("hello", StateFun.stringType());
        let failed = false;
        try {
            TypedValueSupport.parseTypedValue(box, StateFun.floatType());
        } catch (e) {
            failed = true;
        }
        expect(failed).toStrictEqual(true);
    })
});

File: statefun-sdk-python/tests/types_test.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2014 The Apache Software Foundation

import unittest

import statefun
from statefun.utils import to_typed_value


class TypeNameTestCase(unittest.TestCase):

    def assertRoundTrip(self, tpe, value):
        serializer = tpe.serializer()
        out = serializer.serialize(value)
        got = serializer.deserialize(out)
        self.assertEqual(got, value)

    def test_built_ins(self):
        self.assertRoundTrip(statefun.BoolType, True)
        self.assertRoundTrip(statefun.IntType, 0)
        self.assertRoundTrip(statefun.FloatType, float(0.5))
        self.assertRoundTrip(statefun.DoubleType, 1e-20)
        self.assertRoundTrip(statefun.LongType, 1 << 45)
        self.assertRoundTrip(statefun.StringType, "hello world")

    def test_json_type(self):
        import json
        tpe = statefun.simple_type(typename="org.foo.bar/UserJson",
                                   serialize_fn=json.dumps,
                                   deserialize_fn=json.loads)

        self.assertRoundTrip(tpe, {"name": "bob", "last": "mop"})

    def test_message(self):
        typed_value = to_typed_value(statefun.StringType, "hello world")
        msg = statefun.Message(target_typename="foo/bar", target_id="1", typed_value=typed_value)

        self.assertTrue(msg.is_string())
        self.assertEqual(msg.as_string(), "hello world")
